### ISSUE TITLE
Cannot cross-compile ARM64 / aarch64 for Python 3

### ISSUE THREAD

**Issue by lansing** [catboost/catboost#1857]
Problem: "System" Python seems to only support Python 2.7 for aarch64, and cannot figure out how to use my own Python to target a specific version.
catboost version: HEAD
Operating System: Ubuntu 20
CPU: Intel i7
GPU: Not using CUDA

I would like to cross-compile catboost Python package against Python 3.7 but have not been able to get anything working. 

I've tried to cross-compile catboost for aarch64 using `-DUSE_SYSTEM_PYTHON=3.7` (also tried 3.6, 3.8, etc), based on the suggestions from this issue: https://github.com/catboost/catboost/issues/1787
```
max@carcel ~/C/c/c/p/catboost ((9a18764d…)) [1]> ../../../ya make -r -DHAVE_CUDA=no --target-platform default-linux-aarch64 -DUSE_SYSTEM_PYTHON=3.6
Error[-WUserErr]: in $S/build/platform/python/ya.make:104:94: No system Python 3.6 for the selected platform
Configure error (use -k to proceed)
```

This seems to be due to this line:
https://github.com/catboost/catboost/blob/3c3696d0f2d96beba8c8bd8d33659fb1e1c5725e/build/platform/python/ya.make#L13

So, does that imply that we cannot use the `ya make` Python to build for aarch64, if we want to use Python > 2.7?

Is it possible to cross-compile against local Python and other headers? I get an error when I try:

```
max@carcel ~/C/c/c/p/catboost ((9a18764d…)) [1]> ../../../ya make -r -DHAVE_CUDA=no --target-platform default-linux-aarch64 -DPYTHON_CONFIG=/home/max/.pyenv/versions/3.7.10/bin/python3-config -DUSE_ARCADIA_PYTHON=no -DOS_SDK=local
```

Produces:

```
------- [CC] {default-linux-aarch64, pic, FAILED} $(S)/catboost/libs/column_description/feature_tag.cpp
command /home/max/.ya/tools/v3/2343316036/bin/clang++ --target=aarch64-linux-gnu -fsigned-char -march=armv8a -B$OS_SDK_ROOT_RESOURCE_GLOBAL/usr/bin -c -o /home/max/.ya/build/build_root/wv0r/000008/catboost/libs/column_description/feature_tag.cpp.pic.o -I/home/max/.ya/build/build_root/wv0r/000008 -I/home/max/Code/catboost -I/home/max/Code/catboost/contrib/libs/linux-headers -I/home/max/Code/catboost/contrib/libs/linux-headers/_nf -I/home/max/Code/catboost/contrib/libs/cxxsupp/libcxx/include -I/home/max/Code/catboost/contrib/libs/cxxsupp/libcxxrt -I/home/max/Code/catboost/contrib/libs/zlib/include -I/home/max/Code/catboost/contrib/libs/double-conversion/include -I/home/max/Code/catboost/contrib/libs/tbb/include -fdebug-prefix-map=/home/max/.ya/build/build_root/wv0r/000008=/-B -Xclang -fdebug-compilation-dir -Xclang /tmp -pipe -O3 -g -ggnu-pubnames -fexceptions -fcolor-diagnostics -faligned-allocation -fcommon -ffunction-sections -fdata-sections -W -Wall -Wno-parentheses -Wno-unknown-warning-option -Wno-implicit-const-int-float-conversion -Werror -DFAKEID=8510746 -DARCADIA_ROOT=/home/max/Code/catboost -DARCADIA_BUILD_ROOT=/home/max/.ya/build/build_root/wv0r/000008 -D_THREAD_SAFE -D_PTHREADS -D_REENTRANT -D_LIBCPP_ENABLE_CXX17_REMOVED_FEATURES -D_LARGEFILE_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DNDEBUG -D__LONG_LONG_SUPPORTED -DCATBOOST_OPENSOURCE=yes -D_libunwind_ -nostdinc++ -fPIC -fPIC -std=c++20 -Woverloaded-virtual -Wno-invalid-offsetof -Wno-attributes -Wno-dynamic-exception-spec -Wno-register -Wimport-preprocessor-directive-pedantic -Wno-exceptions -Wno-inconsistent-missing-override -Wno-undefined-var-template -Wno-return-std-move -Wno-address-of-packed-member -Wno-defaulted-function-deleted -Wno-enum-compare-switch -Wno-c99-designator -Wno-deprecated-copy -Wno-final-dtor-non-final-class -Wno-initializer-overrides -Wno-pessimizing-move -Wno-range-loop-construct -Wno-reorder-init-list -Wno-unused-lambda-capture -Wno-deprecated-anon-enum-enum-conversion -Wno-deprecated-enum-enum-conversion -Wno-deprecated-enum-float-conversion -Wno-ambiguous-reversed-operator -Wno-deprecated-volatile -DCATBOOST_OPENSOURCE=yes -nostdinc++ /home/max/Code/catboost/catboost/libs/column_description/feature_tag.cpp failed with exit code 1 in /home/max/.ya/build/build_root/wv0r/000008
In file included from /home/max/Code/catboost/catboost/libs/column_description/feature_tag.cpp:1:
In file included from /home/max/Code/catboost/catboost/libs/column_description/feature_tag.h:3:
In file included from /home/max/Code/catboost/util/generic/vector.h:6:
In file included from /home/max/Code/catboost/util/memory/alloc.h:3:
In file included from /home/max/Code/catboost/contrib/libs/cxxsupp/libcxx/include/memory:670:
/home/max/Code/catboost/contrib/libs/cxxsupp/libcxx/include/typeinfo:219:14: error: cast from pointer to smaller type 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') loses information
      return reinterpret_cast<__type_name_t>(__v);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/max/Code/catboost/contrib/libs/cxxsupp/libcxx/include/typeinfo:225:16: error: reinterpret_cast from 'std::__type_info_implementations::__non_unique_arm_rtti_bit_impl::__type_name_t' (aka 'unsigned int') to 'size_t' (aka 'unsigned long') is not allowed
        return reinterpret_cast<size_t>(__v);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Thanks for looking at this. I would love to be able to get a Python package for ARM supporting Python 3.7.

**Comment by kizill** [comment]
I've tried without `-DOS_SDK=local` and it just worked
```bash
python3 ya make -r -DHAVE_CUDA=no --target-platform default-linux-aarch64 catboost/python-package/catboost -DPYTHON_CONFIG=/home/kirillovs/anaconda3/bin/python3-config
```
Can you please try it?

**Comment by kizill** [comment]
Wow, that's not what i was expecting either :( Maybe python-config wasn't fully informative in that case. Can you please also set up paths to python includes and libpython3.X.so by setting -DPYTHON_INCLUDE="-I/home/kirillovs/anaconda3/include/python3.8" -DPYTHON_LIBRARIES="/home/kirillovs/anaconda3/lib/libpython3.8.so"

Sad to say, but currenlty i don't have a fully configured arm platform with linux to test this binaries on :( What are you using in your case, please share your experience :)

**Comment by lansing** [comment]
@kizill I tried with your extra build options, and got the same result.

Here is the build command I used:
```
../../../ya make -r -DHAVE_CUDA=no --target-platform default-linux-aarch64 -DPYTHON_CONFIG=/home/max/.pyenv/versions/3.7.10/bin/python3-config -DPYTHON_INCLUDE="-I/home/max/.pyenv/versions/3.7.10/include/python3.7m/" -DPYTHON_LIBRARIES="/home/max/.pyenv/versions/3.7.10/lib/libpython3.7m.a"
```

Then I built the `whl` and copied it to my test environment. Ultimately, when calling `import catboost` from python 3.7, I get:

```
ImportError: /usr/local/lib64/python3.7/site-packages/catboost/_catboost.so: undefined symbol: Py_EnterRecursiveCall
```

Which is the same error I've gotten before, and which does not occur when I try using the build in Python 3.9.

For testing the build, I'm using Docker with arm64 emulation support.

Here's the Dockerfile I use to build the image I'm testing with (this is a simplified version of our python 3.7 production environment):
```
FROM amazoncorretto:8
RUN yum install -y curl tar gzip procps python3 gcc gcc-c++ python3-devel git zip
```

Then build an arm64 image with that using something like:
```
docker buildx build --platform linux/arm64  -t catboost-build-arm --load -f Dockerfile.python37 .
```

Then I start a container with a mounted directory containing the catboost package I built.

**Comment by xeonvs** [comment]
Do you have any news about how to build a library for the python version != 3.9?

**Comment by zhen-jia** [comment]
Any updates on this? I got the same error as @lansing when using `Python 3.7`.  After changing to `Python3.9`, I got a new error: `_catboost.so: undefined symbol: PyDescr_IsData` Any suggestions? Thanks!

**Comment by arestifo** [comment]
@lansing How did you build the wheel? Using `mk_wheel.py` seems to not have any support for `linux/aarch64` at the minute. I also tried looking into `pip wheel` but that doesn't seem to support cross-complilation or specifying the platform you would like to build the wheel for.

**Comment by lansing** [comment]
@arestifo I'm looking over my old build environment to try and figure this out... it was about a year ago so I don't remember exact details.

It looks like I modified `mk_wheel.py` to add support for `linux_aarch64`. 

Then I ran a command something like this:

```
python mk_wheel.py --target-platform=default-linux-aarch64 -DHAVE_CUDA=no --build-widget=no -DPYTHON_CONFIG=/home/max/.pyenv/versions/3.7.10/bin/python3-config
```

Note that the code to this script has changed since I ran this last year so who knows if this will still work.

Also note that a wheel is just a zip file. You should be able to build the wheel for x86 and then replace the `.so` file (and rename the wheel/zip file to the appropriate architecture).

**Comment by arestifo** [comment]
@lansing Thanks, this is really useful information! I was able to modify `mk_wheel.py` and get the wheel built under the same Python 3.7 environment you describe, but importing it in a Python 3.9 environment results in a new error: `ImportError: dynamic module does not define module export function (PyInit__catboost)`. Have you seen this error before?

I tried importing it into a Python 3.8 and 3.7 environment and got the same error. Definitely a frustrating issue, since this is the only package blocking my team's transition to `linux/arm64` without emulation
